cpp/test/distance/sparse_l2_nn.cu

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (c) 2021-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../test_utils.h"
+#include <gtest/gtest.h>
+#include <raft/distance/detail/sparse_l2_nn.cuh>
+#include <raft/distance/sparse_l2_nn.cuh>
+#include <raft/linalg/norm.cuh>
+#include <raft/random/rng.cuh>
+#include <raft/util/cuda_utils.cuh>
+#include <raft/util/cudart_utils.hpp>
+
+#include <iostream>
+#include <stdio.h>
+
+namespace raft {
+namespace distance {
+namespace sparse_l2_nn {
+
+template <typename LabelT, typename DataT>
+struct CubKVPMinReduce {
```

Comment:
Are you able to use the `raft::KeyValuePair` here? I understand some of the cub functions are only overloaded internally for the cub version but it would be nice to use the raft variant wherever it's possible.